### DISCUSSION TITLE
qdrant indexing

### DISCUSSION THREAD

**Discussion by maaly77** [qdrant/qdrant#3842] [General]
Hi, 
I am trying to index 1.5 M sentences with their vectors from a mistral model as well as normalizing the embeddings. I am indexing it on my server using your docker image. The indexing is taking so long 7 days now and I am not sure if it should take that long or am I doing something wrong. 
Thanks

**Comment by generall** [comment]
Do you mean that mistral embeddings + vector indexing together is taking more than 7 days?

**Reply by maaly77** [reply]
yes, the indexing of 1.5M sentences vectorised by mistral embedding model is taking more than 7 days. I used https://python.langchain.com/docs/integrations/vectorstores/qdrant#on-premise-server-deployment

**Reply by generall** [reply]
do you have an estimation of which fraction of time takes vectorization and which indexing?

**Reply by maaly77** [reply]
it seems that I am running out of memory while indexing. I am using this langchain function to index 5 million sentences with their metadata. It worked with 10 sentences but once I started indexing the 5 million it takes really long time (2 weeks) and the script keep running while encountering memory issues. My server is 128 GB memory. I am currently using this script in a loop where texts are fraction of the 5 million, not sure if it will work?

qdrant = Qdrant.from_documents(
        texts,
        embeddings,
        url=url,
        api_key=None,
        prefer_grpc=True,
        collection_name="mycollection",
        #force_recreate=True,
    )

**Reply by timvisee** [reply]
You may want to disable indexing while doing bulk upserts, which could significantly increase overall performance and indexing time. Please see our documentation here: https://qdrant.tech/documentation/tutorials/bulk-upload/

You likely want to upload to disk directly too, which is also described on that page.

**Reply by maaly77** [reply]
I am using langchain to create and index vectors https://qdrant.tech/documentation/frameworks/langchain/ using mistral embeddings model (On-premise server deployment). So are you implying that this approach can't be used with large datasets? and I will need to create vectors first then upload then index ....??? There is no fast way to do that with langchain framework?

**Comment by maaly77** [comment]
I am not sure how to know the indexing time. I managed to vectorize those sentences in chromadb and they took half day and the total size was 6 GB. Here are screenshots of the structure I had for qdrant collection. Not sure where to look to answer your question
<img width="1186" alt="image" src="https://github.com/qdrant/qdrant/assets/105817666/82f23bf2-3133-458a-8b21-881df1b724a9">
<img width="1182" alt="image" src="https://github.com/qdrant/qdrant/assets/105817666/7f10dc14-2102-44f1-b651-8e70db6d6f6f">

**Comment by minakshimathpal** [comment]
@maaly77 Did you manage to get a solution to this issue. I am struggling with the same issue. "https://qdrant.tech/documentation/tutorials/bulk-upload/" says to upload the vectors first and then update the collection but even after doing that "indexed_vector_count" show zero inside my collection. @timvisee Any views/input on this please

**Reply by maaly77** [reply]
@minakshimathpal I haven't tried the above approach yet. I still using langchain to vectorize and index small collections and querying them and so far working on small datasets. Haven't move to large one yet

**Reply by timvisee** [reply]
How many vectors do you have? You need a certain number of them for any to be indexed.

If you do have a large number of vectors you either didn't set the new indexing threshold correctly (either not set or very high), or it is still ongoing.

You can check the current collection status. If the indexing threshold listed there is not what you expect, you didn't update it properly. If it is, and the collection status is not green, it is ongoing indexing.

**Reply by maaly77** [reply]
@timvisee @minakshimathpal I have given it a try and you are right I get 0 vectors even on small amount of document. I have even tested this https://qdrant.tech/documentation/quickstart/ and I got:

```
Collection name: test_collection
Points count: 6
Vectors count: None
Indexed vectors count: 0
Segments count: 8
An error occurred: object of type 'NoneType' has no len()
```

Not sure what is going on. I can only manage to index using langchain
i am using local qdrant

**Reply by timvisee** [reply]
Qdrant local mode does not support indexes and is bare bones on purpose.

The documentation that suggests to enable indexing later is for use with Qdrant server.

You can get an instance free of charge over at our cloud: https://cloud.qdrant.io/

**Reply by maaly77** [reply]
But I managed to index vectors locally using langchain library
Collection name: x
Points count: 45431
Vectors count: None
Indexed vectors count: 40448
Segments count: 8
An error occurred: 'dict' object has no attribute 'size'

**Reply by timvisee** [reply]
Could you elaborate on how you connect to, and set up, Qdrant through LangChain?

**Reply by maaly77** [reply]
@timvisee I am using this approach https://qdrant.tech/documentation/frameworks/langchain/
On-premise server deployment

**Reply by kirianguiller** [reply]
> @timvisee Qdrant local mode does not support indexes and is bare bones on purpose.
> 
> The documentation that suggests to enable indexing later is for use with Qdrant server.
> 
> You can get an instance free of charge over at our cloud: https://cloud.qdrant.io/



So it is practically difficult to use Qdrant local mode for database having more than millions of embeddings?

I

**Reply by maaly77** [reply]
it is very slow with langchain given we can't use the batch mode but it is doable. it also depends on the model size and context length.
